### ISSUE TITLE
plugins API ?

### ISSUE THREAD

**Issue by ShMcK** [rematch/rematch#8]
We should create a plugins API under init. This can allow for merging store middleware, setting up models, etc.

Plugins may be:
A) an array of objects

```js
init({
  plugins: [plugin(), plugin2()]
})
```

B) an array of strings that internally references an existing local package

```js
init({
  plugins: ['plugin1', 'plugin2']
})
```

C) Another possible API to consider, as used by DVA:

```js
const app = init()
// sets up plugin
app.use(plugin())
```

Regardless, we should decide on a plugin API with some reasoning.

Some things to consider:

1) The plugins are more efficient if they run on startup, which leads me towards A & B.
2) If we are to hold all packages locally, I'd lean towards B. If users are able to create their own packages, or add config settings to a package I'd go with A.

Personally, I'm leaning towards A. What are your thoughts?

**Comment by blairbodnar** [comment]
Yep, I lean towards "A".

The "A" approach also makes it easy to pass config options to the plugin if the plugin has options.

```js
import { init } from 'rematch'
import AwesomePlugin from 'rematch-awesome-plugin'
import RadPlugin from 'rematch-rad-plugin'

init({
  plugins: [
    AwesomePlugin({ coolStuff: true }),
    RadPlugin()
  ],
})
```

**Comment by ShMcK** [comment]
Plugins API still requires more thought.

What might plugins want to change? Some ideas:

- [ ] add models *(example: loading)*
- [ ] add middleware *(example: react-router)*
- [ ] create pre/post hooks of actions or effects *(example: loading)*

Any other ideas of what a plugin might want to effect? It may help by starting about what kind of plugins might be possible.

**Comment by ShMcK** [comment]
After much discussion and experimentation with @blairbodnar ...

The goal is to have the user experience look like this:

```js
import { init, model, dispatch, select } from 'rematch'
```

The plugins API needs 3 methods:

- `onInit` - runs first, can setup exposed values
- `onModel` takes three params:
    - `model`
    - `config` - an internal config that can be mutated
    - `exports` - the final exposed values. Note that they need to already be named exports in order for this to work.
 - `middleware` - new middleware added to the store middleware list

This should be enough to create:
- dispatch
- effects
- selectors
- hooks
- whatever else comes to the party unannounced and is like, "yo, me too"

Essentially these will be created internally with an `init` like setup within `index.js`, that should be the same as the exposed `init`.

**Comment by blairbodnar** [comment]
Got it working! Pretty close to what we were discussing in the comment above. I veered off the path a little here and there but I think the general design is in the right direction. We should be able to change it pretty easy I think. I'm sure you'll have some good ideas to improve it! 

* [https://github.com/rematch/rematch/tree/feature/plugins](https://github.com/rematch/rematch/tree/feature/plugins)
* Everything is a plugin! `dispatch`, `effects`, `selectors`, and `hooks` 👏🏻👏🏻👏🏻👏🏻
* Temporarily removed a bunch of the tests - We can bring them back though! I just did some in plugins.test.js and dispatch.test.js for now. 
* Got lazy and just exported the ugly way into `pluginExports` for now. It shouldn't be a problem to set up the nice named exports though! 
* It builds ok! I built it and tested it out with `create-react-app` and `react-redux`. `dispatch`, `effects`, `selectors`, and `hooks` all work! Here's proof! 😎 [https://www.youtube.com/watch?v=NvOrI0fG_18](https://www.youtube.com/watch?v=NvOrI0fG_18)
* Brain is dead. I go to bed. Will explain more later, I'm sure I'm forgetting something.
